### PR TITLE
145 fix

### LINKED ISSUE #145: [[ClassName]] parsing breaks if pulled in from @inheritdoc
### What steps will reproduce the problem?

1. Put a doc block on a class method/property, which includes a reference to a `[[ClassName]]`
2. Extend the class with another, and override the super method/property. Put `@inheritdoc` in its docblock.

When `@inheritdoc` is replaced with the parent class’ documentation, the doc generator will not  remember which class `ClassName` resolved to in the parent class, unless `ClassName` happens to already be imported in the child class (or in the same namespace).

#### Real World Example

The doc block for `yii\base\Component::behaviors()` references `[[Behavior]]` on [line 427](https://github.com/yiisoft/yii2/blob/9bd7dfc89ed92ddbb156c7fa167c7f53e0188115/framework/base/Component.php#L427). That method is extended by `yii\rest\Controller::behaviors()`, which pulls in the parent docs with an `@inheritdoc` tag on [line 46](https://github.com/yiisoft/yii2/blob/9bd7dfc89ed92ddbb156c7fa167c7f53e0188115/framework/rest/Controller.php#L46).

In the class reference for [`yii\base\Component::behaviors()`](http://www.yiiframework.com/doc-2.0/yii-base-component.html#behaviors()-detail), `[[Behavior]]` got correctly resolved to `yii\base\Behavior`, however the class reference for [`yii\rest\Component::behaviors()`](http://www.yiiframework.com/doc-2.0/yii-rest-controller.html#behaviors()-detail) didn’t know what to do with it, so gave it a `<span class="broken-link">` and incorrectly assumed the fully qualified class name was `\yii\rest\Behavior`.

### What's expected?

In the above example, the doc generator should be smart enough to know that `[[Behavior]]` referenced `yii\base\Behavior` in the class that the documentation was actually written in, and use that.

Since the markdown parsing happens at a different stage, the easiest way to fix this would probably just be to find the closest parent class where `Behavior` actually resolves, and not worry about whether there even was an `@inheritdoc` tag in the first place, etc.

### PR DESCRIPTION
**Pull request by brandonkelly**
This PR fixes #145 where API links weren’t always getting parsed correctly when pulled into a subclass’ property/method via an `@inheritdoc` tag.

My strategy for fixing is:

1. Keep track of where methods & properties were inheriting their docs from with new `$inheritsFrom` properties on MethodDoc & PropertyDoc (9822a90234275be84c0ab63cf282d0770ca1f800)
2. Give `ApiMarkdown::process()` the ability to parse API links across multiple contexts (b3819c8f29a1ecd758a5cfa9d2e31c85957c08ac)
3. Start passing in the method/property `inheritsFrom` values to `ApiMarkdown::process()`, so if any API links can't be resolved from the main context, `inheritsFrom` will be checked as a fallback. (a4542293031bc4e74614d5ba4e28f0d98765009b)

Tested against the Yii 2 source, and this fixes 33 broken API links (108 down to 75). I checked a couple of the remaining broken links, and they were actual doc issues, so not something we can fix here.

This also fixes another bug, where broken links were getting the passed-in context prepended to them, which is a bad assumption. We have no way of knowing what namespace/class name should be added to a broken reference, so the reference should really just be output verbatim, wrapped in `<span class="broken-link">`.

(Originally I attempted a simpler fix, where instead of steps 2 and 3, I just passed the `inheritsFrom` context to `ApiMarkdown::process()` if it was set. However that didn’t work for things like `yii\behaviors\BlameableBehavior::getValue()`, which combines an `@inheritdoc` tag **and** an extended method description that contains its own `[[defaultValue]]` link, which is only defined in the subclass. So the only way to solve stuff like that is to pass multiple contexts.)

| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #145

### CODE REVIEW

**Comment by brandonkelly** [comment]
It occurred to me that there was still a potential to lose track of inherited contexts, if going through multiple recursive levels of @inheritdoc tags, for example:

```php
class Foo
{
    protected $foo;

    /** Returns [[foo]] */
    public setFoo() {
        // ...
    }
}

class Bar extends Foo
{
    protected $bar;

    /** {@inheritdoc} and [[bar]] */
    public setFoo($foo) {
        // ...
    }
}

class Bar extends Baz
{
    protected $baz;

    /** {@inheritdoc} and [[baz]] */
    public setFoo($foo) {
        // ...
    }
}
```

No cases where this would have been an issue in Yii, but guessing we’ll run into it in Craft. So just pushed a fix.

**Comment by brandonkelly** [comment]
Just pushed a drastically simpler fix for this, which doesn’t touch Context.php or any of the template files. I realized that `ApiMarkdownTrait::parseApiLinks()` could assemble its own array of contexts to check, made up of the one passed-in `$context`’s ancestor classes.

File: helpers/ApiMarkdownTrait.php [hunk 1]
```diff
@@ -69,6 +64,38 @@ protected function parseApiLinks($text)
...
+     * @param TypeDoc $type
+     * @param array $contexts
+     * @since 2.1.3
+     */
+    private function _findContexts($type, &$contexts = array())
+    {
+        $contexts[] = $type;
+
+        if ($type instanceof ClassDoc) {
+            if (!empty($type->traits)) {
```
Comment (xepozz) [inline helpers/ApiMarkdownTrait.php]: `if` + `foreach` should be extracted to a separate method to reuse it below in 82 and 91 lines

Comment (rob006) [reply helpers/ApiMarkdownTrait.php]: I would rather remove this `if` - `foreach` will handle empty array just fine.

Comment (brandonkelly) [reply helpers/ApiMarkdownTrait.php]: @rob006 Done: cb74fc722cf86e1b7e4c83568943f18b9d9ec798

**Comment by samdark** [comment]
Thank you!

**Pull request merged**